{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "951f97bb-b50b-4c4e-812d-fbb92049e438",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using device: cuda\n",
      "GPU: NVIDIA A800-SXM4-80GB\n",
      "CPU: Intel(R) Xeon(R) Platinum 8462Y+\n"
     ]
    }
   ],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "import os\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import torch\n",
    "import torch.nn as nn\n",
    "import torch.optim as optim\n",
    "from tqdm.notebook import tqdm\n",
    "import matplotlib.pyplot as plt\n",
    "import time\n",
    "\n",
    "# Evaluation metrics\n",
    "from sklearn.metrics import (homogeneity_score,v_measure_score,adjusted_mutual_info_score,normalized_mutual_info_score,adjusted_rand_score,fowlkes_mallows_score)\n",
    "\n",
    "# Import SpaLP modules    Python package: https://github.com/dbjzs/SpaLP\n",
    "from SpaLP.LP import SpatialLocalPooling\n",
    "from SpaLP.utils import set_seed, prepare_inputs\n",
    "# Set random seed\n",
    "set_seed(7)\n",
    "\n",
    "#Check the GPU memory and CPU memory\n",
    "import memory_profiler  # pip install memory_profiler\n",
    "import torch\n",
    "import gc\n",
    "\n",
    "# Check device\n",
    "import cpuinfo   # pip install py-cpuinfo\n",
    "device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')\n",
    "print(f\"Using device: {device}\")\n",
    "if device.type == 'cuda':\n",
    "    print(f\"GPU: {torch.cuda.get_device_name(0)}\")\n",
    "info = cpuinfo.get_cpu_info()\n",
    "print(\"CPU:\", info['brand_raw'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b987f4c9-617d-4ec4-a44e-6a3ccdb4cba7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Check the GPU memory and CPU memory\n",
    "from memory_profiler import memory_usage\n",
    "def measure_resources(func):\n",
    "    def wrapper(*args, **kwargs):\n",
    "        gc.collect()\n",
    "        torch.cuda.synchronize()\n",
    "        torch.cuda.reset_peak_memory_stats()\n",
    "        torch.cuda.empty_cache()\n",
    "        \n",
    "        before = memory_usage(-1, interval=0.1, timeout=1)[0]\n",
    "        mem_peak, result = memory_usage((func, args, kwargs),max_usage=True,retval=True)\n",
    "        print(f\"CPU peak memory: {(mem_peak - before)/1024:.2f} GB\")\n",
    "        \n",
    "        torch.cuda.synchronize()\n",
    "        gpu_peak = torch.cuda.max_memory_allocated() / 1024**3\n",
    "        print(f\"GPU peak memory: {gpu_peak:.2f} GB\")\n",
    "        return result\n",
    "    return wrapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ee0f0bd0-2274-4600-8909-c10cf85d0dac",
   "metadata": {},
   "outputs": [],
   "source": [
    "def train_spalp(graph, in_channels, hidden_dim, epochs, lr, device,seed):\n",
    "    \"\"\"Train SpaLP model and return embeddings.\"\"\"\n",
    "    from tqdm.auto import tqdm\n",
    "    set_seed(seed)\n",
    "    pbar = tqdm(range(epochs), desc=\"Training\",ncols=200)\n",
    "    model = SpatialLocalPooling(in_channels, hidden_dim).to(device)\n",
    "    \n",
    "    optimizer = optim.Adam(model.parameters(), lr=lr)\n",
    "    criterion = nn.MSELoss()\n",
    "    model.train()\n",
    "    losses = []\n",
    "    \n",
    "    for epoch in pbar:\n",
    "        start_time = time.time()\n",
    "        model.train()\n",
    "        optimizer.zero_grad()\n",
    "        reconstructed,embedding = model(graph.features, graph.neighbor_idx)\n",
    "        loss = criterion(reconstructed, graph.features)\n",
    "        loss.backward()\n",
    "        optimizer.step()\n",
    "        losses.append(loss.item())\n",
    "        elapsed = time.time() - start_time\n",
    "        pbar.set_postfix({\"Epoch\": epoch,\"Loss\": f\"{loss.item():.4f}\"})\n",
    "    \n",
    "    # Get final embeddings\n",
    "    model.eval()\n",
    "    with torch.no_grad():\n",
    "        reconstructed,embedding = model(graph.features, graph.neighbor_idx)\n",
    "        reconstructed = reconstructed.cpu().numpy()\n",
    "        embedding=embedding.cpu().numpy()\n",
    "    return embedding, losses"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b07854f-0d6f-4d12-ad6d-1ed39bd52e36",
   "metadata": {},
   "source": [
    "# Dataset 1: Low quality CosMx Human Kidney Cancer data (1,236,281 cells)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d96c0fbd-3af2-411d-b800-fe25b8bc81f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shape: 1,236,281 cells x 1,000 genes\n",
      "obsm keys: ['spatial']\n"
     ]
    }
   ],
   "source": [
    "# Load CosMx Human Kidney Cancer data\n",
    "adata = sc.read_h5ad('/home/dbj/SpaLA/CosMxKidneycancer/cosmx_kidney.h5ad')\n",
    "adata.var_names_make_unique()\n",
    "print(f\"Shape: {adata.shape[0]:,} cells x {adata.shape[1]:,} genes\")\n",
    "print(f\"obsm keys: {list(adata.obsm.keys())}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e1bea680-d8f6-438c-ad32-285a13c3bafe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Spatial coords shape: (1236281, 2)\n",
      "Preprocessing standardization: 8.85s\n"
     ]
    }
   ],
   "source": [
    "# Preprocessing standardization  *******\n",
    "start_time = time.time()\n",
    "sc.pp.normalize_total(adata, inplace=True)\n",
    "sc.pp.log1p(adata)\n",
    "sc.pp.scale(adata)\n",
    "adata.obsm['feat']=adata.X\n",
    "prep_time = time.time() - start_time\n",
    "\n",
    "print(f\"Spatial coords shape: {adata.obsm['spatial'].shape}\")\n",
    "print(f\"Preprocessing standardization: {prep_time:.2f}s\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "07d286f1-cd1d-4d5c-a789-614cf8100f81",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Feature dimensions: 1000\n"
     ]
    }
   ],
   "source": [
    "# Set the hyperparameters  k=4\n",
    "k=4\n",
    "in_channels = adata.obsm['feat'].shape[1]\n",
    "lr=0.001\n",
    "epochs = 200\n",
    "hidden_dim=64\n",
    "print(f\"Feature dimensions: {in_channels}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "46046485-4f84-4e65-b5fb-0266d344f415",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Graph preparation: 1.56s\n"
     ]
    }
   ],
   "source": [
    "# Build neighbor graph  k=4\n",
    "start_time = time.time()\n",
    "graph = prepare_inputs(adata, k=k,device=device)\n",
    "prep_time = time.time() - start_time\n",
    "print(f\"Graph preparation: {prep_time:.2f}s\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "00f191b5-2716-4e4d-b425-a63c3729519a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Training: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 200/200 [01:03<00:00,  3.15it/s, Epoch=199, Loss=0.9836]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU peak memory: 4.85 GB\n",
      "GPU peak memory: 30.15 GB\n",
      "Training time: 67.80s\n",
      "CPU times: user 1min 3s, sys: 2.85 s, total: 1min 6s\n",
      "Wall time: 1min 7s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "#runtime /GPU peak memory/CPU peak memory\n",
    "start_time = time.time()\n",
    "@measure_resources\n",
    "def run_train():\n",
    "    return train_spalp(graph, in_channels, hidden_dim, epochs, lr, device, seed=7)#The default random seed for the SpaLP algorithm is 7.\n",
    "embedding, losses = run_train()\n",
    "train_time = time.time() - start_time\n",
    "print(f\"Training time: {train_time:.2f}s\")\n",
    "\n",
    "# Store embeddings\n",
    "adata.obsm['SpaLP'] = embedding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e99e68f5-a736-416d-aad0-1ec41784b507",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOStream.flush timed out\n",
      "IOStream.flush timed out\n",
      "IOStream.flush timed out\n",
      "IOStream.flush timed out\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Leiden time: 150min 27.31s\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOStream.flush timed out\n",
      "IOStream.flush timed out\n",
      "IOStream.flush timed out\n",
      "IOStream.flush timed out\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UMAP time: 16min 16.79s\n",
      "Found 7 clusters\n",
      "CPU times: user 6h 13min 7s, sys: 2h 26min 10s, total: 8h 39min 18s\n",
      "Wall time: 2h 46min 44s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Clustering and UMAP\n",
    "start_time = time.time()\n",
    "sc.pp.neighbors(adata, use_rep=\"SpaLP\")\n",
    "sc.tl.leiden(adata, resolution=0.2, random_state=2024, key_added=\"SpaLP\")#The clustering results save in adata.obs['SpaLP']\n",
    "elapsed = time.time() - start_time\n",
    "print(f\"Leiden time: {int(elapsed//60)}min {elapsed%60:.2f}s\")\n",
    "\n",
    "t0 = time.time()\n",
    "sc.tl.umap(adata)\n",
    "elapsed = time.time() - t0\n",
    "print(f\"UMAP time: {int(elapsed//60)}min {elapsed%60:.2f}s\")\n",
    "print(f\"Found {len(adata.obs['SpaLP'].unique())} clusters\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "200b04d6-1a68-4fbb-b639-c0bb04c6131a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 58.7 s, sys: 2min 5s, total: 3min 3s\n",
      "Wall time: 3.76 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# KMeans time\n",
    "from sklearn.cluster import KMeans\n",
    "kmeans = KMeans(n_clusters=15, random_state=2024) \n",
    "labels = kmeans.fit_predict(adata.obsm['SpaLP'])\n",
    "adata.obs['Kmeans_SpaLP']=labels.astype(str)  #The clustering results save in adata.obs['Kmeans_SpaLP']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19bf25e5-daf1-4418-bbb3-69ed3c5b014f",
   "metadata": {},
   "source": [
    "## Obtaining the Leiden and UMAP results using rapid single cell (GPU-scanpy)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "230ca119-bdff-41b3-a1e6-fed3621131d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write_h5ad('/home/dbj/SpaLP/gittest/repoduced/result/Fig6Low—quality-CosMx-Human-Kidney-Cancer.h5ad',compression='gzip')  #Save the embedding for rapid single-cell (GPU-scanpy) analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33f595c5-5204-4d75-8a93-8dd3fe20e567",
   "metadata": {},
   "source": [
    "# Dataset 2: Xenium Gastric cancer data (696,314 cells)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "41d389a3-8579-46d5-865b-8c446e5374cb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shape: 696,314 cells x 377 genes\n",
      "obsm keys: ['spatial']\n"
     ]
    }
   ],
   "source": [
    "adata=sc.read_h5ad('/home/dbj/SpaLA/Gastric_cancer/Gastric.h5ad')\n",
    "adata.var_names_make_unique()\n",
    "print(f\"Shape: {adata.shape[0]:,} cells x {adata.shape[1]:,} genes\")\n",
    "print(f\"obsm keys: {list(adata.obsm.keys())}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "38e2fbea-2e6e-489a-a68c-85aba2ef6f42",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Spatial coords shape: (662434, 2)\n",
      "Preprocessing standardization: 2.18s\n"
     ]
    }
   ],
   "source": [
    "# Preprocessing standardization  *******\n",
    "start_time = time.time()\n",
    "sc.pp.filter_cells(adata,min_genes=2)\n",
    "sc.pp.normalize_total(adata,inplace=True)\n",
    "sc.pp.log1p(adata)\n",
    "sc.pp.scale(adata)\n",
    "adata.obsm['feat']=adata.X\n",
    "prep_time = time.time() - start_time\n",
    "\n",
    "print(f\"Spatial coords shape: {adata.obsm['spatial'].shape}\")\n",
    "print(f\"Preprocessing standardization: {prep_time:.2f}s\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b25d2231-84c8-4e33-bd4b-503b13c11f8d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Feature dimensions: 377\n"
     ]
    }
   ],
   "source": [
    "# Set the hyperparameters  k=5\n",
    "k=5\n",
    "in_channels = adata.obsm['feat'].shape[1]\n",
    "lr=0.001\n",
    "epochs = 200\n",
    "hidden_dim=64\n",
    "print(f\"Feature dimensions: {in_channels}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8f61f767-3540-4ab9-a011-17bd647988cf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Graph preparation: 0.56s\n"
     ]
    }
   ],
   "source": [
    "# Build neighbor graph  k=5\n",
    "start_time = time.time()\n",
    "graph = prepare_inputs(adata, k=k,device=device)\n",
    "prep_time = time.time() - start_time\n",
    "print(f\"Graph preparation: {prep_time:.2f}s\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "80fb4c79-328f-4709-aa52-bd0d76259d10",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Training: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 200/200 [00:37<00:00,  5.39it/s, Epoch=199, Loss=0.9581]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU peak memory: 1.27 GB\n",
      "GPU peak memory: 13.41 GB\n",
      "Training time: 38.99s\n",
      "CPU times: user 37.1 s, sys: 919 ms, total: 38.1 s\n",
      "Wall time: 39 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "#runtime /GPU peak memory/CPU peak memory\n",
    "start_time = time.time()\n",
    "@measure_resources\n",
    "def run_train():\n",
    "    return train_spalp(graph, in_channels, hidden_dim, epochs, lr, device, seed=7)#The default random seed for the SpaLP algorithm is 7.\n",
    "embedding, losses = run_train()\n",
    "train_time = time.time() - start_time\n",
    "print(f\"Training time: {train_time:.2f}s\")\n",
    "\n",
    "# Store embeddings\n",
    "adata.obsm['SpaLP'] = embedding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e99dfe87-f40c-4f0d-a648-46f384b124a0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOStream.flush timed out\n",
      "IOStream.flush timed out\n",
      "IOStream.flush timed out\n",
      "IOStream.flush timed out\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Leiden time: 63min 29.93s\n",
      "UMAP time: 7min 22.29s\n",
      "Found 5 clusters\n",
      "CPU times: user 2h 21min 40s, sys: 55min 30s, total: 3h 17min 10s\n",
      "Wall time: 1h 10min 52s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Clustering and UMAP\n",
    "start_time = time.time()\n",
    "sc.pp.neighbors(adata, use_rep=\"SpaLP\")\n",
    "sc.tl.leiden(adata, resolution=0.2, random_state=2024, key_added=\"SpaLP\")#The clustering results save in adata.obs['SpaLP']\n",
    "elapsed = time.time() - start_time\n",
    "print(f\"Leiden time: {int(elapsed//60)}min {elapsed%60:.2f}s\")\n",
    "\n",
    "t0 = time.time()\n",
    "sc.tl.umap(adata)\n",
    "elapsed = time.time() - t0\n",
    "print(f\"UMAP time: {int(elapsed//60)}min {elapsed%60:.2f}s\")\n",
    "print(f\"Found {len(adata.obs['SpaLP'].unique())} clusters\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "62af35b3-f4fb-42be-8cdf-9893d990ac49",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 47.3 s, sys: 1min 18s, total: 2min 6s\n",
      "Wall time: 2.64 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# KMeans time\n",
    "from sklearn.cluster import KMeans\n",
    "kmeans = KMeans(n_clusters=15, random_state=2024) \n",
    "labels = kmeans.fit_predict(adata.obsm['SpaLP'])\n",
    "adata.obs['Kmeans_SpaLP']=labels.astype(str)  #The clustering results save in adata.obs['Kmeans_SpaLP']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a24fd2e-f4a5-40e3-81a7-ca945af15088",
   "metadata": {},
   "source": [
    "## Obtaining the Leiden and UMAP results using rapid single cell (GPU-scanpy)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a8bc3dfe-e719-4884-9d74-2485ddafe65d",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write_h5ad('/home/dbj/SpaLP/gittest/repoduced/result/Fig6Xenium_Gastric_cancer.h5ad',compression='gzip')  #Save the embedding for rapid single-cell (GPU-scanpy) analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f511acaf-db93-45dc-9f15-1f976dc5006d",
   "metadata": {},
   "source": [
    "# Dataset 3: Xenium Milti-omics cancer data (465,545 cells)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "bb6ee9cc-b73e-4e6f-90cf-5a8af5e5e8ac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(AnnData object with n_obs × n_vars = 465545 × 396\n",
       "     obs: 'x_centroid', 'y_centroid', 'transcript_counts', 'control_probe_counts', 'genomic_control_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'nucleus_count', 'segmentation_method', 'TACCO_celltype'\n",
       "     var: 'gene_name', 'feature_type'\n",
       "     obsm: 'spatial',\n",
       " AnnData object with n_obs × n_vars = 465545 × 27\n",
       "     obs: 'x_centroid', 'y_centroid', 'transcript_counts', 'control_probe_counts', 'genomic_control_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'nucleus_count', 'segmentation_method', 'TACCO_celltype'\n",
       "     var: 'gene_name', 'feature_type'\n",
       "     obsm: 'spatial')"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata1=sc.read_h5ad('/home/dbj/SpaLA/Xeniummultiomics/RNA.h5ad')\n",
    "adata2=sc.read_h5ad('/home/dbj/SpaLA/Xeniummultiomics/Protein.h5ad')\n",
    "adata1,adata2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "8919224a-4e6d-4a05-a46a-6654772c476f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Spatial coords shape: (465545, 2)\n",
      "Preprocessing standardization: 2.59s\n"
     ]
    }
   ],
   "source": [
    "# Preprocessing standardization  *******\n",
    "start_time = time.time()\n",
    "sc.pp.normalize_total(adata1,inplace=True)\n",
    "sc.pp.log1p(adata1)\n",
    "sc.pp.scale(adata1)\n",
    "adata1.obsm['feat']=adata1.X\n",
    "sc.pp.normalize_total(adata2,inplace=True)\n",
    "sc.pp.log1p(adata2)\n",
    "sc.pp.scale(adata2)\n",
    "adata2.obsm['feat']=adata2.X\n",
    "\n",
    "adata=sc.AnnData(np.hstack([adata1.obsm['feat'],adata2.obsm['feat']]))\n",
    "sc.pp.scale(adata)\n",
    "adata.var_names = adata1.var_names.tolist() + adata2.var_names.tolist()\n",
    "rna_len = adata1.var_names.shape[0]\n",
    "prot_len = adata2.var_names.shape[0]\n",
    "adata.var['feature_type'] = ['RNA'] * rna_len + ['Protein'] * prot_len\n",
    "adata.obs_names=adata2.obs_names\n",
    "adata.obsm['spatial']=adata2.obsm['spatial']\n",
    "adata.obsm['feat']=adata.X\n",
    "prep_time = time.time() - start_time\n",
    "\n",
    "print(f\"Spatial coords shape: {adata.obsm['spatial'].shape}\")\n",
    "print(f\"Preprocessing standardization: {prep_time:.2f}s\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "4d231511-634e-4fe7-b448-2bda0e1f5fc3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Feature dimensions: 423\n"
     ]
    }
   ],
   "source": [
    "# Set the hyperparameters  k=4\n",
    "k=4\n",
    "in_channels = adata.obsm['feat'].shape[1]\n",
    "lr=0.001\n",
    "epochs = 200\n",
    "hidden_dim=64\n",
    "print(f\"Feature dimensions: {in_channels}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "9aeff630-2ab1-428e-a222-476aceccab40",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Graph preparation: 0.34s\n"
     ]
    }
   ],
   "source": [
    "# Build neighbor graph  k=4\n",
    "start_time = time.time()\n",
    "graph = prepare_inputs(adata, k=k,device=device)\n",
    "prep_time = time.time() - start_time\n",
    "print(f\"Graph preparation: {prep_time:.2f}s\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "d2c2e56b-db0b-459f-822d-b9dbf0f55bf1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Training: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 200/200 [00:27<00:00,  7.27it/s, Epoch=199, Loss=0.9423]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU peak memory: 0.69 GB\n",
      "GPU peak memory: 7.89 GB\n",
      "Training time: 29.70s\n",
      "CPU times: user 28 s, sys: 767 ms, total: 28.7 s\n",
      "Wall time: 29.7 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "#runtime /GPU peak memory/CPU peak memory\n",
    "start_time = time.time()\n",
    "@measure_resources\n",
    "def run_train():\n",
    "    return train_spalp(graph, in_channels, hidden_dim, epochs, lr, device, seed=7)#The default random seed for the SpaLP algorithm is 7.\n",
    "embedding, losses = run_train()\n",
    "train_time = time.time() - start_time\n",
    "print(f\"Training time: {train_time:.2f}s\")\n",
    "\n",
    "# Store embeddings\n",
    "adata.obsm['SpaLP'] = embedding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "60ae5b0b-5c1c-4faf-95f1-0ab4191c55d0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOStream.flush timed out\n",
      "IOStream.flush timed out\n",
      "IOStream.flush timed out\n",
      "IOStream.flush timed out\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Leiden time: 23min 17.02s\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOStream.flush timed out\n",
      "IOStream.flush timed out\n",
      "IOStream.flush timed out\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UMAP time: 4min 37.49s\n",
      "Found 7 clusters\n",
      "CPU times: user 1h 8min, sys: 30min 6s, total: 1h 38min 7s\n",
      "Wall time: 27min 54s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Clustering and UMAP\n",
    "start_time = time.time()\n",
    "sc.pp.neighbors(adata, use_rep=\"SpaLP\")\n",
    "sc.tl.leiden(adata, resolution=0.2, random_state=2024, key_added=\"SpaLP\")#The clustering results save in adata.obs['SpaLP']\n",
    "elapsed = time.time() - start_time\n",
    "print(f\"Leiden time: {int(elapsed//60)}min {elapsed%60:.2f}s\")\n",
    "\n",
    "t0 = time.time()\n",
    "sc.tl.umap(adata)\n",
    "elapsed = time.time() - t0\n",
    "print(f\"UMAP time: {int(elapsed//60)}min {elapsed%60:.2f}s\")\n",
    "print(f\"Found {len(adata.obs['SpaLP'].unique())} clusters\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ad029665-1cdc-42d0-a4c2-22e19754d19f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 43.4 s, sys: 1min 4s, total: 1min 47s\n",
      "Wall time: 2.02 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# KMeans time\n",
    "from sklearn.cluster import KMeans\n",
    "kmeans = KMeans(n_clusters=15, random_state=2024) \n",
    "labels = kmeans.fit_predict(adata.obsm['SpaLP'])\n",
    "adata.obs['Kmeans_SpaLP']=labels.astype(str)  #The clustering results save in adata.obs['Kmeans_SpaLP']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "57a3fd11-cfa0-4a9d-a881-4431bd4b0358",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write_h5ad('/home/dbj/SpaLP/gittest/repoduced/result/Fig6Xenium_multi-omics.h5ad',compression='gzip')  #Save the embedding for rapid single-cell (GPU-scanpy) analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6639b3d-be84-4810-901b-bc0e36aa7835",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:SpaLPgit]",
   "language": "python",
   "name": "conda-env-SpaLPgit-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
